File: android/drengr/src/main/kotlin/dev/drengr/sdk/IngestSink.kt
package dev.drengr.sdk

import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.ArrayDeque
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import kotlin.math.min
import kotlin.random.Random

/**
 * Batches captured signals and ships them to the Drengr ingest endpoint,
 * authenticated by a publishable key. Port of the proven Dart/JS IngestSink,
 * carrying every device-run lesson from birth:
 *  - delivery uses a DEDICATED OkHttpClient with NO Drengr interceptor, so the
 *    sink can never capture its own POSTs (the self-capture loop can't exist);
 *  - the persist scheduler serializes writes on a single-thread executor, so no
 *    starvation loop is possible;
 *  - envelope carries sent_at_ms for server-side clock-skew correction.
 *
 * Best-effort and non-blocking: never throws into the app, drops oldest on
 * overflow, retries with exponential backoff + full jitter, persists the queue
 * to a JSONL file in the app's files dir so an app kill doesn't lose events.
 */
internal class IngestSink(
    filesDir: File,
    private val url: String,
    private val publishableKey: String,
    private val context0: Map<String, Any?>,
    private val maxBatch: Int = 50,
    private val maxQueue: Int = 500,
    private val flushIntervalMs: Long = 10_000,
) {
    // A single-thread executor is the whole concurrency model: enqueue, flush,
    // persist, and restore all run here, so the in-memory queue needs no locks
    // and no write can overlap another (no starvation loop).
    private val exec: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor { r ->
        Thread(r, "drengr-sink").apply { isDaemon = true }
    }
    private val queue = ArrayDeque<JSONObject>()
    private val file = File(filesDir, "drengr_queue.jsonl")
    private var flushScheduled = false
    private var retries = 0

    // Session-scoped identity/experiment state, merged into every envelope (see flush()).
    // Mutated only on `exec` (identify/setExperiment dispatch there), same as the queue.
    private var externalId: String? = null
    private val experiments = LinkedHashMap<String, String>()

    // Delivery client: NO interceptor → structurally invisible to capture.
    private val delivery = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .callTimeout(30, TimeUnit.SECONDS)
        .build()

    init {
        exec.execute { restore() }
    }

    fun addNetwork(e: NetworkEvent) {
        val ev = try { toNet(e) } catch (_: Throwable) { return }
        exec.execute { enqueue(ev) }
    }

    /** Sets external_id (all events hereafter) and emits one identify event; traits
     *  go through the same redact+project pipeline as bodies. Fail-open. */
    fun identify(externalId: String, traits: Map<String, Any?> = emptyMap()) {
        if (externalId.isEmpty()) return
        val redactedTraits = try {
            Redact.projectBody(Redact.redactBody(JSONObject(traits).toString()))
        } catch (_: Throwable) { null }
        val ev = try {
            JSONObject().apply {
                put("kind", "identify")
                put("event_id", randomId())
                put("ts_ms", System.currentTimeMillis())
                put("external_id", externalId)
                redactedTraits?.let { put("traits", it) }
            }
        } catch (_: Throwable) { return }
        exec.execute {
            this.externalId = externalId
            enqueue(ev)
        }
    }

    /** Sets/clears a session-scoped experiment variant (all events hereafter, as
     *  `experiments`). A null/empty variant clears the key. Fail-open. */
    fun setExperiment(key: String, variant: String?) {
        if (key.isEmpty()) return
        exec.execute {
            if (variant.isNullOrEmpty()) experiments.remove(key) else experiments[key] = variant
        }
    }

    private fun toNet(e: NetworkEvent): JSONObject {
        val status = e.statusCode ?: 0
        val failed = e.errorText != null || status >= 400
        val o = JSONObject()
        o.put("kind", if (failed) "net_fail" else "net")
        o.put("event_id", randomId())
        o.put("ts_ms", e.timestampMs)
        o.put("method", e.method)
        o.put("url", e.url)
        o.put("status", status)
        o.put(
            "error_kind",
            if (failed) (if (e.errorText != null) "transport" else if (status >= 500) "server" else "client") else "",
        )
        o.put("duration_ms", e.durationMs)
        o.put("req_bytes", e.requestBodyBytes)
        o.put("resp_bytes", e.responseBodyBytes)
        Redact.projectBody(e.requestBody)?.let { o.put("req_body", it) }
        Redact.projectBody(e.responseBody)?.let { o.put("body", it) }
        return o
    }

    private fun enqueue(ev: JSONObject) {
        queue.addLast(ev)
        while (queue.size > maxQueue) queue.pollFirst()
        schedulePersist()
        if (retries > 0) return
        if (queue.size >= maxBatch) flush() else scheduleFlush()
    }

    private fun scheduleFlush() {
        if (flushScheduled) return
        flushScheduled = true
        exec.schedule({ flushScheduled = false; flush() }, flushIntervalMs, TimeUnit.MILLISECONDS)
    }

    private fun flush() {
        if (queue.isEmpty()) return
        val batch = ArrayList<JSONObject>(min(queue.size, 1000))
        while (queue.isNotEmpty() && batch.size < 1000) batch.add(queue.pollFirst())

        val envelope = JSONObject()
        for ((k, v) in context0) envelope.put(k, v ?: JSONObject.NULL)
        envelope.put("sent_at_ms", System.currentTimeMillis())
        envelope.put("events", JSONArray(batch))
        externalId?.let { envelope.put("external_id", it) }
        if (experiments.isNotEmpty()) envelope.put("experiments", JSONObject(experiments))

        var acked = false
        var permanent = false
        try {
            val req = Request.Builder()
                .url(url)
                .header("authorization", "Bearer $publishableKey")
                .post(envelope.toString().toRequestBody(JSON))
                .build()
            delivery.newCall(req).execute().use { resp ->
                acked = resp.isSuccessful
                // Non-retriable 4xx (revoked key 401, bad batch 400/413) never
                // succeeds — retrying forever head-of-line-blocks the queue. Drop.
                permanent = resp.code in 400..499 && resp.code != 429 && resp.code != 408
            }
        } catch (_: Throwable) {
            acked = false
        }

        if (acked || permanent) {
            retries = 0 // batch consumed (delivered or dropped as permanent)
            schedulePersist()
            if (queue.isNotEmpty()) scheduleFlush()
        } else {
            // Requeue at the front; shed newest on overflow.
            for (i in batch.indices.reversed()) queue.addFirst(batch[i])
            while (queue.size > maxQueue) queue.pollLast()
            schedulePersist()
            armBackoff()
        }
    }

    private fun armBackoff() {
        val exp = BASE_BACKOFF_MS * (1L shl min(retries, 20))
        val capped = min(exp, MAX_BACKOFF_MS)
        val delay = BASE_BACKOFF_MS + Random.nextLong(capped)
        retries++
        exec.schedule({ flush() }, delay, TimeUnit.MILLISECONDS)
    }

    // --- persistence (serialized on exec; writer loops nothing — one write per tick) ---
    private fun schedulePersist() {
        if (flushScheduled) { /* a flush is queued; persist will ride along */ }
        exec.execute { persist() }
    }

    private fun persist() {
        try {
            if (queue.isEmpty()) {
                if (file.exists()) file.delete()
            } else {
                val tmp = File(file.parentFile, file.name + ".tmp")
                tmp.bufferedWriter().use { w ->
                    for (ev in queue) { w.write(ev.toString()); w.newLine() }
                }
                tmp.renameTo(file)
            }
        } catch (_: Throwable) { /* best-effort */ }
    }

    private fun restore() {
        try {
            if (!file.exists()) return
            file.forEachLine { line ->
                if (line.isNotBlank() && queue.size < maxQueue) {
                    try { queue.addLast(JSONObject(line)) } catch (_: Throwable) {}
                }
            }
            if (queue.isNotEmpty()) scheduleFlush()
        } catch (_: Throwable) { /* corrupt/missing → start empty */ }
    }

    private fun randomId(): String {
        val sb = StringBuilder(32)
        repeat(16) { sb.append("%02x".format(Random.nextInt(256))) }
        return sb.toString()
    }

    companion object {
        private const val BASE_BACKOFF_MS = 2_000L
        private const val MAX_BACKOFF_MS = 5 * 60_000L
        private val JSON = "application/json; charset=utf-8".toMediaTypeOrNull()
    }
}

File: ios/Sources/Drengr/IngestSink.swift
import Foundation
import CryptoKit

/// Batches captured signals and ships them to the Drengr ingest endpoint,
/// authenticated by a publishable key. Port of the proven Dart/JS/Kotlin
/// IngestSink, carrying every device-run lesson from birth:
///  - delivery uses a URLSession with NO Drengr protocol, so the sink can never
///    capture its own POSTs (the self-capture loop can't exist);
///  - a serial DispatchQueue is the whole concurrency model — enqueue, flush,
///    persist all run on it, so the queue needs no locks and no write can overlap
///    another (no starvation loop);
///  - the envelope carries sent_at_ms for server-side clock-skew correction.
///
/// Best-effort and non-blocking: never throws into the app, drops oldest on
/// overflow, retries with exponential backoff + full jitter, persists the queue
/// to a JSONL file in Application Support so an app kill doesn't lose events.
final class IngestSink {
    private let url: URL
    private let publishableKey: String
    private var context: [String: Any]   // mutated only on `queue` (session rotation)
    private let maxBatch: Int
    private let maxQueue: Int
    private let flushInterval: TimeInterval

    private let queue = DispatchQueue(label: "dev.drengr.sink")
    private var events: [[String: Any]] = []
    private var flushScheduled = false
    private var retries = 0
    private let fileURL: URL

    // Session-scoped identity/experiment state, merged into every envelope (see
    // flush()). Mutated only on `queue`, same as `events`.
    private var externalId: String?
    private var experimentsMap: [String: String] = [:]

    // On-device sealing config. nil = fail-closed: PII ships as typed placeholders,
    // never raw. Mutated/read only on `queue`.
    private var vault: VaultConfig?

    // Delivery session WITHOUT the Drengr protocol → invisible to capture.
    let delivery: URLSession

    init(url: URL, publishableKey: String, context: [String: Any],
         maxBatch: Int = 50, maxQueue: Int = 500, flushInterval: TimeInterval = 10,
         session: URLSession? = nil) {
        self.url = url
        self.publishableKey = publishableKey
        self.context = context
        self.maxBatch = maxBatch
        self.maxQueue = maxQueue
        self.flushInterval = flushInterval
        self.delivery = session ?? {
            let cfg = URLSessionConfiguration.ephemeral
            cfg.timeoutIntervalForRequest = 30
            return URLSession(configuration: cfg)
        }()
        URLSessionCapture.markInternal(self.delivery)
        let dir = (try? FileManager.default.url(for: .applicationSupportDirectory, in: .userDomainMask,
                                                appropriateFor: nil, create: true))
            ?? FileManager.default.temporaryDirectory
        self.fileURL = dir.appendingPathComponent("drengr_queue.jsonl")
        queue.async { [weak self] in self?.restore() }
    }

    func addNetwork(_ e: NetworkEvent) {
        let (ev, pii) = toNet(e)
        queue.async { [weak self] in self?.ingest(ev, pii: pii) }
    }

    /// Activate (or clear) on-device sealing after the org's signed config verifies.
    /// Until set, the sink is fail-closed: PII ships as placeholders, never raw.
    func setVault(_ v: VaultConfig?) {
        queue.async { [weak self] in self?.vault = v }
    }

    // Seal captured PII into the event when a vault is set, else fail-closed. Mirrors
    // the JS sealAndEnqueue: attach subject_hash + the HPKE envelope as `pii`.
    private func ingest(_ event: [String: Any], pii: [String: String]) {
        var event = event
        if let v = vault, !pii.isEmpty { sealInto(&event, pii: pii, vault: v) }
        enqueue(event)
    }

    private func sealInto(_ event: inout [String: Any], pii: [String: String], vault: VaultConfig) {
        let sh = externalId.map { Identity.subjectHash(pepper: vault.orgPepper, externalId: $0) } ?? ""
        if !sh.isEmpty { event["subject_hash"] = sh }
        let bh = sha256hex((event["req_body"] as? String ?? "") + "|" + (event["body"] as? String ?? ""))
        guard #available(iOS 17, macOS 14, tvOS 17, *) else { return }
        let ctx = SealContext(subjectHash: sh, sealUlid: event["event_id"] as? String ?? "",
                              bodyHash: bh, tenantId: vault.tenantId)
        if let env = Seal.sealPii(pii, ctx: ctx, orgPubKey: vault.orgPubKey, kid: vault.kid) {
            event["pii"] = ["pii_alg": env.pii_alg, "kid": env.kid, "enc": env.enc, "ct": env.ct]
        }
    }

    private func sha256hex(_ s: String) -> String {
        SHA256.hash(data: Data(s.utf8)).map { String(format: "%02x", $0) }.joined()
    }

    /// Enqueue an already-redacted semantic tap (kind "tap" — ingest-allowlisted).
    func addTap(_ e: TapEvent) {
        var o: [String: Any] = [
            "kind": "tap",
            "event_id": randomID(),
            "ts_ms": e.timestampMs,
            "screen": e.screen,
            "label": e.label ?? "",
            "label_source": e.labelSource,
            "role": e.role,
            "element_class": e.elementClass,
            "x": e.x,
            "y": e.y,
        ]
        if let v = e.value { o["value"] = v }
        queue.async { [weak self] in self?.enqueue(o) }
    }

    /// Enqueue a pre-built behavior event (screen_view / rage_tap / dead_tap /
    /// crash — already redacted by the capture layer). Fills event_id/ts_ms
    /// when absent, mirroring the Flutter sink. Fail-open.
    func addBehavior(_ ev: [String: Any]) {
        var o = ev
        if o["event_id"] == nil { o["event_id"] = randomID() }
        if o["ts_ms"] == nil { o["ts_ms"] = Int64(Date().timeIntervalSince1970 * 1000) }
        queue.async { [weak self] in self?.enqueue(o) }
    }

    /// app_foreground: process entered the foreground on `screen`. Lets the
    /// backend subtract backgrounded gaps when computing true screen dwell.
    /// Mirrors Android `IngestSink.addAppForeground`.
    func addAppForeground(_ screen: String) { addLifecycle("app_foreground", screen: screen) }

    /// app_background: process left the foreground on `screen`.
    func addAppBackground(_ screen: String) { addLifecycle("app_background", screen: screen) }

    private func addLifecycle(_ kind: String, screen: String) {
        var o: [String: Any] = [
            "kind": kind,
            "event_id": randomID(),
            "ts_ms": Int64(Date().timeIntervalSince1970 * 1000),
        ]
        if !screen.isEmpty { o["screen"] = screen }
        queue.async { [weak self] in self?.enqueue(o) }
    }

    /// Crash-path enqueue: waits (bounded) until the event is queued AND
    /// persisted to disk, so the imminent process death can't lose it. The
    /// timeout guarantees a wedged queue can never turn a crash into a hang.
    func addCrashSync(_ ev: [String: Any], timeout: TimeInterval = 1.0) {
        var o = ev
        if o["event_id"] == nil { o["event_id"] = randomID() }
        if o["ts_ms"] == nil { o["ts_ms"] = Int64(Date().timeIntervalSince1970 * 1000) }
        let sem = DispatchSemaphore(value: 0)
        queue.async { [weak self] in
            if let self = self {
                self.events.append(o)
                while self.events.count > self.maxQueue { self.events.removeFirst() }
                self.persist()
            }
            sem.signal()
        }
        _ = sem.wait(timeout: .now() + timeout)
    }

    /// Sets external_id (all events hereafter) and emits one identify event;
    /// traits go through the same redact+project pipeline as bodies. Fail-open.
    func identify(_ externalId: String, traits: [String: Any] = [:]) {
        guard !externalId.isEmpty else { return }
        var redactedTraits: String?
        if !traits.isEmpty,
           let data = try? JSONSerialization.data(withJSONObject: traits, options: []),
           let json = String(data: data, encoding: .utf8) {
            redactedTraits = Classify.classifyBody(Redact.redactBody(json)).projection
        }
        var ev: [String: Any] = [
            "kind": "identify",
            "event_id": randomID(),
            "ts_ms": Int64(Date().timeIntervalSince1970 * 1000),
            "external_id": externalId,
        ]
        if let rt = redactedTraits { ev["traits"] = rt }
        queue.async { [weak self] in
            self?.externalId = externalId
            self?.enqueue(ev)
        }
    }

    /// Sets/clears a session-scoped experiment variant (all events hereafter, as
    /// `experiments`). A nil/empty `variant` clears `key`. Fail-open.
    func setExperiment(_ key: String, variant: String?) {
        guard !key.isEmpty else { return }
        queue.async { [weak self] in
            guard let self = self else { return }
            if let v = variant, !v.isEmpty {
                self.experimentsMap[key] = v
            } else {
                self.experimentsMap.removeValue(forKey: key)
            }
        }
    }

    /// Flush the pending queue under the CURRENT (old) session id, then adopt the
    /// new one — rotation never mislabels queued events.
    func rotateSession(to newId: String) {
        queue.async { [weak self] in
            guard let self = self else { return }
            self.flush()   // envelope snapshots the old session_id synchronously
            self.context["session_id"] = newId
        }
    }

    /// Force-send everything queued now, bypassing batch/backoff timers.
    func forceFlush(_ completion: (() -> Void)? = nil) {
        queue.async { [weak self] in
            guard let self = self else { completion?(); return }
            self.retries = 0
            self.flush(completion)
        }
    }

    private func toNet(_ e: NetworkEvent) -> ([String: Any], [String: String]) {
        let status = e.statusCode ?? 0
        let failed = e.errorText != nil || status >= 400
        var o: [String: Any] = [
            "kind": failed ? "net_fail" : "net",
            "event_id": randomID(),
            "ts_ms": e.timestampMs,
            "method": e.method,
            "url": e.url,
            "status": status,
            "error_kind": failed ? (e.errorText != nil ? "transport" : (status >= 500 ? "server" : "client")) : "",
            "duration_ms": e.durationMs,
            "req_bytes": e.requestBodyBytes,
            "resp_bytes": e.responseBodyBytes,
        ]
        // Seal-by-default classifier: free-text PII that value-scrubbing can't
        // pattern-match seals to a typed placeholder in the projection; the raw value
        // goes to piiMap for HPKE sealing (see sealInto). projection is leak-free.
        let reqC = Classify.classifyBody(e.requestBody)
        let respC = Classify.classifyBody(e.responseBody)
        var pii: [String: String] = [:]
        for (k, v) in reqC.piiMap { pii["req." + k] = v }
        for (k, v) in respC.piiMap { pii["resp." + k] = v }
        if let rb = reqC.projection { o["req_body"] = rb }
        if let sb = respC.projection { o["body"] = sb }
        o["req_headers"] = capHeaders(e.requestHeaders)
        o["resp_headers"] = capHeaders(e.responseHeaders)
        return (o, pii)
    }

    /// Cap the already-redacted header map: 48 entries max; oversized serialization → {}.
    private func capHeaders(_ h: [String: String]) -> [String: String] {
        var out = h
        if out.count > Self.headersMaxEntries {
            out = Dictionary(uniqueKeysWithValues:
                h.sorted { $0.key < $1.key }.prefix(Self.headersMaxEntries).map { ($0.key, $0.value) })
        }
        guard let d = try? JSONSerialization.data(withJSONObject: out, options: []) else { return [:] }
        return d.count > Self.headersMaxJsonBytes ? [:] : out
    }

    private static let headersMaxEntries = 48
    private static let headersMaxJsonBytes = 8192

    private func enqueue(_ ev: [String: Any]) {
        events.append(ev)
        while events.count > maxQueue { events.removeFirst() }
        schedulePersist()
        if retries > 0 { return }
        if events.count >= maxBatch { flush() } else { scheduleFlush() }
    }

    private func scheduleFlush() {
        if flushScheduled { return }
        flushScheduled = true
        queue.asyncAfter(deadline: .now() + flushInterval) { [weak self] in
            self?.flushScheduled = false
            self?.flush()
        }
    }

    private func flush(_ completion: (() -> Void)? = nil) {
        if events.isEmpty { completion?(); return }
        let batch = Array(events.prefix(1000))
        events.removeFirst(batch.count)

        var envelope = context
        envelope["sent_at_ms"] = Int64(Date().timeIntervalSince1970 * 1000)
        envelope["events"] = batch
        if let id = externalId { envelope["external_id"] = id }
        if !experimentsMap.isEmpty { envelope["experiments"] = experimentsMap }
        guard let body = try? JSONSerialization.data(withJSONObject: envelope, options: []) else {
            // Unserializable batch: drop it rather than loop forever.
            completion?()
            return
        }

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("Bearer \(publishableKey)", forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = body

        let task = delivery.dataTask(with: req) { [weak self] _, response, _ in
            guard let self = self else { completion?(); return }
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            let acked = (200..<300).contains(code)
            // A 4xx (revoked/invalid key 401, bad batch 400/413, quota 429-excepted)
            // will NEVER succeed on retry — retrying it forever head-of-line-blocks
            // the queue and drops every newer event + drains the battery. DROP the
            // batch on a non-retriable 4xx; only 5xx/network errors retry. 429 is
            // retriable (transient rate-limit).
            let permanent = (400..<500).contains(code) && code != 429 && code != 408
            self.queue.async {
                if acked || permanent {
                    self.retries = 0
                    self.schedulePersist()          // batch consumed (delivered or dropped)
                    if !self.events.isEmpty { self.scheduleFlush() }
                } else {
                    self.events.insert(contentsOf: batch, at: 0)
                    while self.events.count > self.maxQueue { self.events.removeLast() }
                    self.schedulePersist()
                    self.armBackoff()
                }
                completion?()
            }
        }
        URLSessionCapture.markInternal(task)
        task.resume()
    }

    private func armBackoff() {
        let exp = Double(Self.baseBackoffMs) * pow(2, Double(min(retries, 20)))
        let capped = min(exp, Double(Self.maxBackoffMs))
        let delay = (Double(Self.baseBackoffMs) + Double.random(in: 0..<capped)) / 1000.0
        retries += 1
        queue.asyncAfter(deadline: .now() + delay) { [weak self] in self?.flush() }
    }

    // --- persistence (serialized on `queue`; one write per tick, no loop) ---
    private func schedulePersist() {
        queue.async { [weak self] in self?.persist() }
    }

    private func persist() {
        do {
            if events.isEmpty {
                try? FileManager.default.removeItem(at: fileURL)
            } else {
                let lines = events.compactMap { ev -> String? in
                    guard let d = try? JSONSerialization.data(withJSONObject: ev, options: []) else { return nil }
                    return String(data: d, encoding: .utf8)
                }
                let tmp = fileURL.appendingPathExtension("tmp")
                try lines.joined(separator: "\n").write(to: tmp, atomically: true, encoding: .utf8)
                _ = try? FileManager.default.replaceItemAt(fileURL, withItemAt: tmp)
            }
        } catch {
            // best-effort
        }
    }

    private func restore() {
        guard let content = try? String(contentsOf: fileURL, encoding: .utf8) else { return }
        for line in content.split(separator: "\n") {
            if events.count >= maxQueue { break }
            if let d = line.data(using: .utf8),
               let obj = try? JSONSerialization.jsonObject(with: d) as? [String: Any] {
                events.append(obj)
            }
        }
        if !events.isEmpty { scheduleFlush() }
    }

    private func randomID() -> String {
        var s = ""
        for _ in 0..<16 { s += String(format: "%02x", Int.random(in: 0..<256)) }
        return s
    }

    private static let baseBackoffMs = 2_000
    private static let maxBackoffMs = 5 * 60_000
}
